polygeist/include/mlir/Dialect/Polygeist/Utils/TransformUtils.h

```diff
@@ -70,76 +63,59 @@ class LoopVersionCondition {
   LoopVersionCondition(AffineCondition affineCond)
       : versionCondition(affineCond) {}
 
+  bool hasSCFCondition() const {
+    return std::holds_alternative<SCFCondition>(versionCondition);
+  }
   SCFCondition getSCFCondition() const {
-    assert(std::holds_alternative<SCFCondition>(versionCondition) &&
-           "expecting valid SCF condition");
+    assert(hasSCFCondition() && "expecting valid SCF condition");
     return std::get<SCFCondition>(versionCondition);
   }
+
+  bool hasAffineCondition() const {
+    return std::holds_alternative<AffineCondition>(versionCondition);
+  }
   AffineCondition getAffineCondition() const {
-    assert(std::holds_alternative<AffineCondition>(versionCondition) &&
-           "expecting valid affine condition");
+    assert(hasAffineCondition() && "expecting valid affine condition");
     return std::get<AffineCondition>(versionCondition);
   }
 
 private:
   std::variant<SCFCondition, AffineCondition> versionCondition;
 };
 
-/// Abstract base class to version a loop like operation.
-class LoopVersionBuilder : public IfThenElseBodyInterface {
+/// Version a loop like operation.
+class LoopVersionBuilder {
 public:
-  static std::unique_ptr<LoopVersionBuilder> create(LoopLikeOpInterface);
+  LoopVersionBuilder(LoopLikeOpInterface loop) : loop(loop) {}
 
-  virtual void versionLoop(const LoopVersionCondition &) const;
+  void versionLoop(const LoopVersionCondition &) const;
 
 protected:
-  LoopVersionBuilder(LoopLikeOpInterface loop) : loop(loop) {}
-
-  void versionLoop(RegionBranchOpInterface) const;
+  void createElseBody(scf::IfOp) const;
+  void createElseBody(AffineIfOp) const;
 
   mutable LoopLikeOpInterface loop; /// The loop to version.
 };
 
-/// Concrete class to version a SCF loop operation.
```

Comment:
Removed derived classes - now `LoopVersionBuilder` can version either SCF loops or affine loops using the `LoopVersionCondition` supplied.